### ISSUE TITLE
Vite js server keep running after stopping or restarting app

### ISSUE THREAD

**Issue by TanguyPa** [Eptagone/Vite.AspNetCore#1]
I'm having an issue with the package usage.
When I start my ASP .NET Core MVC project by setting the middleware to autorun vite dev server, the middleware works fine. 
But when I stop my project or restart it, the vite dev server instance is not stopped. 
So I end up with several instances of the vite server.
Would it be possible when stopping the asp core server that the vite server stops. 

Thank you for your work on this package, it allows to integrate quickly and easily to mvc projects

**Comment by Eptagone** [comment]
Hello friend.
Hmm. That's strange. The vite command is supposed to be closed since it's part of the same instance.

Are you using the Ctrl + 5 and F5 commands together while using Visual Studio IDE? Both processes run separately, so all other terminals that Visual Studio opens must be closed in order for the vite dev server instance to close.

If this is not your case, please provide more details to try to reproduce it.

**Comment by Eptagone** [comment]
Hello @TanguyPa. It seems that [this problem is something common](https://github.com/dotnet/aspnetcore/issues/5204) when starting node processes.
It may not be able to be fixed soon. Anyway, I'll keep checking it to see if I find a solution.
In the meantime, [avoid using IIS](https://github.com/dotnet/aspnetcore/issues/5204#issuecomment-581085135) or start Dev Server externally, without AutoRun option.

**Comment by khalidabuhakmeh** [comment]
My guess is the AppPool in IIS never truly "shuts down" so the parent process is maintaining a connection to the Vite Server and keeping it alive.

I think if this project wants to support IIS, it might need to hook into the IIS lifecycle events.

**Comment by henriksen** [comment]
I get a similar issue when running from the command line with `dotnet watch run`. If I restart the app with `Ctrl+R` a time or two the Vite dev server is not always closed. Sometimes it is, sometimes it isn't. Have yet to see a pattern, but it might not be directly related to IIS hosting.

**Comment by austins** [comment]
Also seeing this issue on Windows and also with dotnet watch.

**Comment by khalidabuhakmeh** [comment]
How are you folks running `dotnet watch` ?

Try running it in `dotnet watch --no-hot-reload` as that will close the process down and then restart it. I'll give it a try and see if I can recreate the issue.

**Comment by khalidabuhakmeh** [comment]
👋 Hi, folks. I found this line in Oleksii Holub's (@Tyrrrz) CliWrap project. 

I can try setting `CreateNoWindow` to `true` to see if that helps. I would guess that the child process might be attached to the host console (Hot Reload), and thus it doesn't exit. 

It's worth testing out, but not sure if it will solve the Windows issues.

https://github.com/Tyrrrz/CliWrap/blob/3eb76065612d38543b19aae37cb0e2b85bd7d8dc/CliWrap/Command.Execution.cs#LL80C1-L85C34

**Comment by khalidabuhakmeh** [comment]
I just submitted a PR and tested it on Windows.

I was able to recreate the issue with IIS Express, but could not track down the issue. 

As an interim fix, the code now checks to see if the port is occupied, if it is, it will close the old process and start a new one. This should work in general, including in scenarios with `dotnet watch`.

**Comment by Eptagone** [comment]
The main problem is that the vite script keeps running until you close the console. It happens when the project is run using IIS, dotnet watch and generally from the Visual Studio IDE.

I just tested The PR #18 from @khalidabuhakmeh. His solution might work provisionally, but I have my doubts about killing the port because what if that port is used by a process other than Vite? Wouldn't it be a trouble if we just kill it?

**Comment by Eptagone** [comment]
The React template for SPA projects runs the React development server in a second console, and that process does stop if the project does too. I haven't figured out how it works yet but I think the full solution might be close that way.

**Comment by khalidabuhakmeh** [comment]
@Eptagone I agree with your points, you didn't have to merge my PR. 😅

But I want to address your points from another perspective.

> What if that port is used by a process other than Vite?
 
Another process might use the port, but it's the port the user has in their configuration/setup. We can assume that they picked it with intentionality. 

> Wouldn't it be trouble if we just kill it?

It had it coming. THIS IS OUR PORT! (just kidding). The user can probably restart those processes, and knowing there is a conflict, they can change their environment. This is development, after all.

We may want to add in the `README` that IIS Express and .NET Watch might work in a way that doesn't kill the parent process, so folks using those environments might want to consider not using the auto-start functionality or know that we'll restart those processes for them.

**Comment by Eptagone** [comment]
Ups. Well, anyway, I think it may be a temporary solution.
I'll add an option like `KillUsedPort` so the user can choose if he wants to kill the busy port.

**Comment by austins** [comment]
I recently experimented with launching an npm script (in my case, for pnpm and webpack via `pnpm run dev`) in a project.

I made a singleton class based on Microsoft's [SpaProxyLaunchManager](https://github.com/dotnet/aspnetcore/blob/1121b2bbb123ad9044d593955e7a2ef863fcf2e5/src/Middleware/Spa/SpaProxy/src/SpaProxyLaunchManager.cs). They have a bit of comments in there hinting that they've also run into quirks with shutting down the process.

Namely, the stop script for Windows/Mac and disposal methods seem to do the trick. It seems to run quite well for me so far even with `dotnet watch`.

Sharing a gist here in case you find any bit of code that might be useful: https://gist.github.com/austins/6b44c48ca724e21f5fea2d549ffd7619
